### PR TITLE
Check to ensure the extracted program enforces effect linearity

### PR DESCRIPTION
**Pull request by yihozhang**
* tests for linearity check

### CODE REVIEW

File: dag_in_context/src/linearity.rs [hunk 1]
```diff
@@ -167,4 +169,118 @@ impl<'a> Extractor<'a> {
...
+                    return Err("There are unconsumed effectful operators".to_string());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Expr {
+    pub fn collect_reachable(
```
Comment (oflatt) [inline dag_in_context/src/linearity.rs:240]: Add a comment- populates reachable_from with ...

File: dag_in_context/src/linearity.rs [hunk 2]
```diff
@@ -167,4 +169,118 @@ impl<'a> Extractor<'a> {
...
+                    return Some(rcexpr);
+                }
+                None
+            };
+
+            for (region, exprs) in reachables {
+                let mut dangling_effectful: HashSet<*const Expr> = exprs
+                    .iter()
+                    .filter_map(|&expr| {
+                        // This is because from_raw creates a new Rc
```
Comment (oflatt) [inline dag_in_context/src/linearity.rs:197]: out of date comment

File: dag_in_context/src/linearity.rs [hunk 3]
```diff
@@ -167,4 +169,118 @@ impl<'a> Extractor<'a> {
...
+                        return Some(rcexpr);
+                    }
+                } else if this.is_effectful(rcexpr) {
+                    return Some(rcexpr);
+                }
+                None
+            };
+
+            for (region, exprs) in reachables {
+                let mut dangling_effectful: HashSet<*const Expr> = exprs
```
Comment (oflatt) [inline dag_in_context/src/linearity.rs:194]: add comment- get all of the expressions, these need to get used

**Review by oflatt — CHANGES_REQUESTED** [review]
awesome work, I'm very happy with this checker
had a few nits

File: dag_in_context/src/linearity.rs (hunk 2 above)
Comment (yihozhang) [reply dag_in_context/src/linearity.rs:197]: good catch!

**Comment by yihozhang** [comment]
All comments should be resolved.

**Pull request merged**